### PR TITLE
Person and group created at for any order ingestion

### LINKED ISSUE #7454: Person created_at / first seen handling
Currently created_at is used either as event timestamp or as current timestamp, this is confusing and wouldn't work well when we do migrations for example.

See https://github.com/PostHog/posthog/pull/7109#discussion_r758363070

Ideally upon ingestion we update this value always to the oldest timestamp we see tied to this person, this would make things work properly for out of order ingestion too. This likely won't increase the number of updates/complexity by much, but should be taken into account.

### PR DESCRIPTION
**Pull request by tiina303**
## Changes

Closes https://github.com/PostHog/posthog/issues/7454

## How did you test this code?

CI

### CODE REVIEW

File: plugin-server/src/worker/ingestion/properties-updater.ts [hunk 1]
```diff
@@ -115,19 +105,32 @@ export async function upsertGroup(
             }
 
             if (propertiesUpdate.updated) {
-                // :TRICKY: insertGroup will raise a RaceConditionError if group was inserted in-between fetch and this
-                const upsertMethod = group ? 'updateGroup' : 'insertGroup'
```
Comment (tiina303) [inline plugin-server/src/worker/ingestion/properties-updater.ts:119]: while this is clever it makes it hard to use code editing/discovery tools effectively, e.g. find references

File: plugin-server/src/worker/ingestion/properties-updater.ts [hunk 2]
```diff
@@ -98,7 +88,7 @@ export async function upsertGroup(
             const group: Group | undefined = await db.fetchGroup(teamId, groupTypeIndex, groupKey, client, {
                 forUpdate: true,
             })
-            const createdAt = group?.created_at || timestamp
+            const createdAt = DateTime.min(group?.created_at || DateTime.now(), timestamp)
```
Comment (yakkomajuri) [inline plugin-server/src/worker/ingestion/properties-updater.ts:91]: Is `group?.created_at` of type `DateTime`?

File: plugin-server/src/worker/ingestion/properties-updater.ts (hunk 1 above)
Comment (yakkomajuri) [reply plugin-server/src/worker/ingestion/properties-updater.ts:119]: let's then have those functions accept an object and do something like:

```js
const groupUpdateArgs = {
    teamId,
    groupTypeIndex,
    groupKey,
    createdAt,
    version,
    client,
    properties: propertiesUpdate.properties,
    propertiesLastUpdatedAt: propertiesUpdate.properties_last_updated_at,
    propertiesLastOperation: propertiesUpdate.properties_last_operation,
}

if (group) {
    await db.updateGroup(groupUpdateArgs)
} else {
    // :TRICKY: insertGroup will raise a RaceConditionError if group was inserted in-between fetch and this
    await db.insertGroup(groupUpdateArgs)
}
```

File: plugin-server/tests/clickhouse/postgres-parity.test.ts [hunk 3]
```diff
@@ -64,8 +64,9 @@ describe('postgres parity', () => {
         const uuid = new UUIDT().toString()
         const person = await hub.db.createPerson(
             DateTime.utc(),
-            { userProp: 'propValue' },
-            { userPropOnce: 'propOnceValue' },
+            { userPropOnce: 'propOnceValue', userProp: 'propValue' },
+            { userProp: DateTime.now().toString(), userPropOnce: DateTime.now().toString() },
```
Comment (yakkomajuri) [inline plugin-server/tests/clickhouse/postgres-parity.test.ts:68]: nit: would be best to set these to the same constant value rather

File: plugin-server/tests/clickhouse/postgres-parity.test.ts [hunk 4]
```diff
@@ -160,7 +165,10 @@ describe('postgres parity', () => {
         // update date and boolean to false
 
         const randomDate = DateTime.utc().minus(100000).setZone('UTC')
```
Comment (yakkomajuri) [inline plugin-server/tests/clickhouse/postgres-parity.test.ts:167]: nit: outside scope here but this doesn't look random to me :D

File: plugin-server/tests/shared/process-event.ts [hunk 5]
```diff
@@ -376,6 +380,7 @@ export const createProcessEventTests = (
             const chPeople = await hub.db.fetchPersons(Database.ClickHouse)
             expect(chPeople.length).toEqual(1)
             expect(JSON.parse(chPeople[0].properties)).toEqual(expectedProps)
+            expect(chPeople[0].created_at).toEqual(now.toFormat('yyyy-MM-dd HH:mm:ss.000'))
```
Comment (yakkomajuri) [inline plugin-server/tests/shared/process-event.ts:383]: nit: hmm not sure how readable this is

File: plugin-server/src/utils/db/db.ts [hunk 6]
```diff
@@ -599,6 +592,41 @@ export class DB {
...
+                JSON.stringify(propertiesLastOperation),
+                personId,
+            ],
+            'updatePersonProperties',
+            client
+        )
+
+        if (updateResult.rows.length === 0) {
+            // this function should always be called in a transaction with person fetch locking for update before
+            throw new RaceConditionError('Failed updating person properties, person was not locked properly')
```
Comment (yakkomajuri) [inline plugin-server/src/utils/db/db.ts:625]: is this really the only reason we wouldn't update?

**Review by yakkomajuri — COMMENTED** [review]
looking good overall, a few comments

File: plugin-server/src/worker/ingestion/properties-updater.ts (hunk 1 above)
Comment (tiina303) [reply plugin-server/src/worker/ingestion/properties-updater.ts:119]: this actually would increase total lines of code due to updateGroup & insertGroup needing to have the input defined differently + in the future we might want to change them independently for params, so thinking of not changing it.

File: plugin-server/src/worker/ingestion/properties-updater.ts (hunk 2 above)
Comment (tiina303) [reply plugin-server/src/worker/ingestion/properties-updater.ts:91]: yes based on types here: https://github.com/PostHog/posthog/blob/master/plugin-server/src/types.ts#L533

File: plugin-server/tests/shared/process-event.ts (hunk 5 above)
Comment (tiina303) [reply plugin-server/tests/shared/process-event.ts:383]: it's not really, but I couldn't figure out a nicer way to do it.

**Comment by tiina303** [comment]
Some worker/ingestion tests are failing not sure if this pr or from before

**Review by yakkomajuri — APPROVED** [review]
overall lgtm, let's make sure tests pass

**Pull request merged**